### PR TITLE
Minor refactor of DeleteCollection in preparation for pagination support there

### PR DESCRIPTION
**Pull request by wojtek-t**
Ref https://github.com/kubernetes/kubernetes/issues/90743

```release-note
NONE
```

/kind cleanup
/sig api-machinery
/priority important-longterm

/assign @tkashem

### CODE REVIEW

**Review by tkashem — COMMENTED** [review]
not strictly within the scope of this refactor, but i thought i would mention the following

File: staging/src/k8s.io/apiserver/pkg/registry/generic/registry/store.go [hunk 1]
```diff
@@ -1212,10 +1190,30 @@ func (e *Store) DeleteCollection(ctx context.Context, deleteValidation rest.Vali
...
 		}()
 	}
-	wg.Wait()
-	// notify distributor to exit
-	close(workersExited)
-	<-distributorExited
+	// In case of all workers exit, notify distributor.
+	go func() {
+		defer utilruntime.HandleCrash(func(panicReason interface{}) {
+			errs <- fmt.Errorf("DeleteCollection distributor panicked: %v", panicReason)
```
Comment (Argh4k) [inline staging/src/k8s.io/apiserver/pkg/registry/generic/registry/store.go:1196]: Is this actually the distributor that panicked?

Comment (wojtek-t) [reply staging/src/k8s.io/apiserver/pkg/registry/generic/registry/store.go]: fair - fixed

File: staging/src/k8s.io/apiserver/pkg/registry/generic/registry/store.go [hunk 2]
```diff
@@ -1212,10 +1190,30 @@ func (e *Store) DeleteCollection(ctx context.Context, deleteValidation rest.Vali
...
+		})
+		wg.Wait()
+		close(workersExited)
+	}()
+
+	func() {
```
[the remark this answers is not in the document]
Comment (wojtek-t) [reply staging/src/k8s.io/apiserver/pkg/registry/generic/registry/store.go:1202]: done

File: staging/src/k8s.io/apiserver/pkg/registry/generic/registry/store.go [hunk 3]
```diff
@@ -1212,10 +1190,30 @@ func (e *Store) DeleteCollection(ctx context.Context, deleteValidation rest.Vali
...
+		for i :=0; i < len(items); i++ {
+			select {
+			case toProcess<-i:
+			case <-workersExited:
```
[the remark this answers is not in the document]
Comment (wojtek-t) [reply staging/src/k8s.io/apiserver/pkg/registry/generic/registry/store.go:1208]: That will kind-of happen implicitly.
To be more specific - as soon as context is closed, the requests in workers will start failing and all workers will exit (which will trigger this case).
There is a test that is actually exercising it:
https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiserver/pkg/registry/generic/registry/store_test.go#L2158

**Review by wojtek-t — COMMENTED** [review]
@tkashem - PTAL

**Comment by tkashem** [comment]
/lgtm

Thanks!

**Pull request merged**